tests/unit/test_charm.py

```diff
@@ -15,49 +13,54 @@
 
 APP_NAME = "canonical-livepatch-server-k8s"
 
+TEST_TOKEN = "test-token"  # nosec
+
 
 class MockOutput:
+    """A wrapper class for command output and errors."""
+
     def __init__(self, stdout, stderr):
         self._stdout = stdout
         self._stderr = stderr
 
     def wait_output(self):
+        """return the stdout and stderr from running the command."""
         return self._stdout, self._stderr
 
 
-def mock_exec(_, command, environment):
+def mock_exec(_, command, environment) -> MockOutput:
+    """Mock Execute the commands."""
     if len(command) != 1:
         return MockOutput("", "unexpected number of commands")
-    if command[0] == "/usr/bin/pg_isready":
+    cmd: str = command[0]
+    if cmd == "/usr/bin/pg_isready":
         return MockOutput(0, "")
-    elif command[0] == "/usr/local/bin/livepatch-schema-tool upgrade /usr/src/livepatch/schema-upgrades":
+    if cmd == "/usr/local/bin/livepatch-schema-tool upgrade /usr/src/livepatch/schema-upgrades":
         return MockOutput("", "")
     return MockOutput("", "unexpected command")
 
 
 class TestCharm(unittest.TestCase):
+    """A wrapper class for charm unit tests."""
+
     def setUp(self):
         self.harness = Harness(LivepatchCharm)
         self.addCleanup(self.harness.cleanup)
         self.harness.disable_hooks()
         self.harness.add_oci_resource("livepatch-server-image")
         self.harness.add_oci_resource("livepatch-schema-upgrade-tool-image")
         self.harness.begin()
-
-        self.tempdir = tempfile.TemporaryDirectory()
-        self.addCleanup(self.tempdir.cleanup)
-        self.harness.charm.framework.charm_dir = pathlib.Path(self.tempdir.name)
-
+        rel_id = self.harness.add_relation("livepatch", "livepatch")
```

Comment:
Glad you were able to figure this one out, apologies for not looking sooner.

Reply:
Oh I see, we actually just moved the add relation from the individual tests into the setUp

Reply:
Thank you Kian, it turned out to be a very good exercise and I learned a lot.

Reply:
the container_pebble call required the relation, which was not created before that call, so this fixed it.